src/controller/CHIPDeviceController.cpp

```diff
@@ -264,54 +278,61 @@ void DeviceController::OnLocalNOCChainGeneration(void * context, CHIP_ERROR stat
+    ChipLogProgress(Controller, "Generating new local credentials");
+    mOperationalKey.Initialize();
 
-    ReturnErrorCodeIf(keypair == nullptr, CHIP_ERROR_NO_MEMORY);
+    chip::Platform::ScopedMemoryBuffer<uint8_t> CSR;
+    size_t csrLength = kMaxCHIPCSRLength;
+    ReturnErrorCodeIf(!CSR.Alloc(csrLength), CHIP_ERROR_NO_MEMORY);
 
-    if (!fabric->AreCredentialsAvailable())
-    {
-        chip::Platform::ScopedMemoryBuffer<uint8_t> CSR;
-        size_t csrLength = kMaxCHIPCSRLength;
-        ReturnErrorCodeIf(!CSR.Alloc(csrLength), CHIP_ERROR_NO_MEMORY);
+    ReturnErrorOnFailure(mOperationalKey.NewCertificateSigningRequest(CSR.Get(), csrLength));
 
-        ReturnErrorOnFailure(keypair->NewCertificateSigningRequest(CSR.Get(), csrLength));
+    ChipLogProgress(Controller, "Getting certificate chain for the controller from the issuer");
 
-        ChipLogProgress(Controller, "Getting certificate chain for the controller from the issuer");
+    // As per specifications section 11.22.5.1. Constant RESP_MAX
+    constexpr uint16_t kMaxRspLen = 900;
+    chip::Platform::ScopedMemoryBuffer<uint8_t> csrElements;
+    ReturnErrorCodeIf(!csrElements.Alloc(kMaxRspLen), CHIP_ERROR_NO_MEMORY);
 
-        // As per specifications section 11.22.5.1. Constant RESP_MAX
-        constexpr uint16_t kMaxRspLen = 900;
-        chip::Platform::ScopedMemoryBuffer<uint8_t> csrElements;
-        ReturnErrorCodeIf(!csrElements.Alloc(kMaxRspLen), CHIP_ERROR_NO_MEMORY);
+    TLV::TLVWriter csrElementWriter;
+    TLV::TLVType containerType;
+    csrElementWriter.Init(csrElements.Get(), kMaxRspLen);
+    ReturnErrorOnFailure(csrElementWriter.StartContainer(TLV::AnonymousTag, TLV::TLVType::kTLVType_Structure, containerType));
+    ReturnErrorOnFailure(csrElementWriter.Put(TLV::ContextTag(1), ByteSpan(CSR.Get(), csrLength)));
+
+    // TODO - Need a mechanism to generate CSRNonce for commissioner's CSR
+    ReturnErrorOnFailure(csrElementWriter.Put(TLV::ContextTag(2), ByteSpan()));
+    ReturnErrorOnFailure(csrElementWriter.Put(TLV::ContextTag(3), ByteSpan()));
+    ReturnErrorOnFailure(csrElementWriter.Put(TLV::ContextTag(4), ByteSpan()));
+    ReturnErrorOnFailure(csrElementWriter.Put(TLV::ContextTag(5), ByteSpan()));
+    ReturnErrorOnFailure(csrElementWriter.EndContainer(containerType));
+    ReturnErrorOnFailure(csrElementWriter.Finalize());
+
+    mOperationalCredentialsDelegate->SetNodeIdForNextNOCRequest(mLocalDeviceId);
+    mOperationalCredentialsDelegate->SetFabricIdForNextNOCRequest(0);
 
-        TLV::TLVWriter csrElementWriter;
-        TLV::TLVType containerType;
-        csrElementWriter.Init(csrElements.Get(), kMaxRspLen);
-        ReturnErrorOnFailure(csrElementWriter.StartContainer(TLV::AnonymousTag, TLV::TLVType::kTLVType_Structure, containerType));
-        ReturnErrorOnFailure(csrElementWriter.Put(TLV::ContextTag(1), ByteSpan(CSR.Get(), csrLength)));
+    // TODO - Need a mechanism to generate signature for commissioner's CSR
```

Comment:
No, since the commissioner will have direct trusted access to its CA and the means to obtain operational credentials should not rely on the CSR flows as in the common commissionable nodes case

Reply:
Agreed. This will get removed when we move commissioner's credentials generation out of CHIPDeviceController.cpp to commissioner application code.